base/iddict.jl

```diff
@@ -83,18 +83,39 @@ function sizehint!(d::IdDict, newsz)
     rehash!(d, newsz)
 end
 
-function setindex!(d::IdDict{K,V}, @nospecialize(val), @nospecialize(key)) where {K, V}
+@inline function ht_keyindex!(d::IdDict{K, V}, @nospecialize(key)) where {K, V}
     !isa(key, K) && throw(ArgumentError("$(limitrepr(key)) is not a valid key for type $K"))
-    if !(val isa V) # avoid a dynamic call
-        val = convert(V, val)::V
-    end
+
+    ht = d.ht
+    t = @_gc_preserve_begin ht
+
+    ref = RefValue{Ptr{Memory{Any}}}(pointer_from_objref(ht))
+    # # keyindex - where a key is stored, or -pos if the key was not present and was inserted at pos
+    keyindex = ccall(:jl_table_assign_bp, Cssize_t, (Ptr{Ptr{Memory{Any}}}, Any, Any, Cint), ref, key, C_NULL, 0)
+    d.ht = unsafe_pointer_to_objref(ref[])::Memory{Any}
```

Comment:
This is a violation of the memory model as the value is not rooted (and cannot be rooted again after pointer_from_objref destroys that information chain)